File: Source/Runtime/CSharp/Halak.Bibim.Runtime/UI/Events/UIEventID.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Halak.Bibim.UI.Events
{
    public struct UIEventID
    {
        #region TypeCode
        public enum TypeCode
        {
            Unknown = 0,
            KeyDown = 1,
            KeyUp = 2,
            KeyPressing = 3,
            MouseEnter = 4,
            MouseLeave = 5,
            MouseMove = 6,
            MouseButtonDown = 7,
            MouseButtonUp = 8,
            MouseButtonPressing = 9,
            MouseWheel = 10,
            MouseClick = 11,
            GamePadButtonDown = 12,
            GamePadButtonUp = 13,
            GamePadButtonPressing = 14,
            GamePadTrigger = 15,
            GamePadThumbstick = 16,
            Application = 17,
        }
        #endregion

        public static readonly UIEventID Unknown;
        public static readonly UIEventID KeyDown;
        public static readonly UIEventID KeyUp;
        public static readonly UIEventID KeyPressing;
        public static readonly UIEventID MouseEnter;
        public static readonly UIEventID MouseLeave;
        public static readonly UIEventID MouseMove;
        public static readonly UIEventID MouseButtonDown;
        public static readonly UIEventID MouseButtonUp;
        public static readonly UIEventID MouseButtonPressing;
        public static readonly UIEventID MouseWheel;
        public static readonly UIEventID MouseClick;
        public static readonly UIEventID GamePadButtonDown;
        public static readonly UIEventID GamePadButtonUp;
        public static readonly UIEventID GamePadButtonPressing;
        public static readonly UIEventID GamePadTrigger;
        public static readonly UIEventID GamePadThumbstick;
    }
}

File: Source/Runtime/CPlusplus/UI/Bibim/UIEventID.cpp
#include <Bibim/PCH.h>
#include <Bibim/UIEventID.h>

namespace Bibim
{
    const String UIEventID::EventNames[ApplicationType] = 
    {
        "Unknown",
        "KeyDown",
        "KeyUp",
        "KeyPressing",
        "MouseEnter",
        "MouseLeave",
        "MouseMove",
        "MouseButtonDown",
        "MouseButtonUp",
        "MouseButtonPressing",
        "MouseWheel",
        "MouseClick",
        "GamePadButtonDown",
        "GamePadButtonUp",
        "GamePadButtonPressing",
        "GamePadTrigger",
        "GamePadThumbstick",
   };

    const UIEventID UIEventID::Unknown(UnknownType);
    const UIEventID UIEventID::KeyDown(KeyDownType);
    const UIEventID UIEventID::KeyUp(KeyUpType);
    const UIEventID UIEventID::KeyPressing(KeyPressingType);
    const UIEventID UIEventID::MouseEnter(MouseEnterType);
    const UIEventID UIEventID::MouseLeave(MouseLeaveType);
    const UIEventID UIEventID::MouseMove(MouseMoveType);
    const UIEventID UIEventID::MouseButtonDown(MouseButtonDownType);
    const UIEventID UIEventID::MouseButtonUp(MouseButtonUpType);
    const UIEventID UIEventID::MouseButtonPressing(MouseButtonPressingType);
    const UIEventID UIEventID::MouseWheel(MouseWheelType);
    const UIEventID UIEventID::MouseClick(MouseClickType);
    const UIEventID UIEventID::GamePadButtonDown(GamePadButtonDownType);
    const UIEventID UIEventID::GamePadButtonUp(GamePadButtonUpType);
    const UIEventID UIEventID::GamePadButtonPressing(GamePadButtonPressingType);
    const UIEventID UIEventID::GamePadTrigger(GamePadTriggerType);
    const UIEventID UIEventID::GamePadThumbstick(GamePadThumbstickType);

    const String& UIEventID::ConvertFromTypeToName(Type type)
    {
        if (type != ApplicationType)
            return EventNames[type];
        else
            return String::Empty;
    }

    UIEventID::Type UIEventID::ConvertFromNameToType(const String& name)
    {
        if (name.IsEmpty() == false)
        {
            for (int i = 0; i < sizeof(EventNames) / sizeof(EventNames[0]); i++)
            {
                if (EventNames[i].EqualsIgnoreCase(name))
                    return static_cast<Type>(i);
            }

            static const String click = "click";
            static const String wheel = "wheel";
            static const String mouseDown = "mousedown";
            static const String mouseUp = "mouseup";
            static const String mousePressing = "mousepressing";

            if (name.EqualsIgnoreCase(click))
                return MouseClickType;
            else if (name.EqualsIgnoreCase(wheel))
                return MouseWheelType;
            else if (name.EqualsIgnoreCase(mouseDown))
                return MouseButtonDownType;
            else if (name.EqualsIgnoreCase(mouseUp))
                return MouseButtonUpType;
            else if (name.EqualsIgnoreCase(mousePressing))
                return MouseButtonPressingType;
        }

        return ApplicationType;
    }
}